### PR TITLE
refactor(*): favor optional chaining over lodash.get

### PR DESCRIPTION
**Pull request by anotherchrisberry**
Lodash's `get` method is extremely convenient. However, it can mask structural errors that Typescript is good at checking against. For example:

```typescript
interface IBar {
  bazzes: string[];
}

interface IFoo {
  bar?: IBar;
}
const foo: IFoo = {
  bar: {
    bars: ['a', 'b']
  }
}

const lodashedThings = get(foo, 'baz.bazzes', []) // [];
const guardedThings = foo?.baz.bazzes ?? []; // throws a TS compile error
```

There are hundreds of places in the code where we can add better type checking by taking advantage of optional chaining — and, to some extent, nullish coalescing (`??`). This is just to get it into the codebase a bit.

We don't want to replace _all_ uses of `get` — it's useful in some places, like form validation, where we embedded dot-separated paths into configuration for fields. But most of the time, we're using it because we never had a better tool.

### CODE REVIEW

File: app/scripts/modules/amazon/src/securityGroup/configure/ingressRuleGroupSelector.component.ts [hunk 1]
```diff
@@ -29,7 +35,7 @@ class IngressRuleSelectorController implements IController {
   public crossAccountAccounts: IAccountDetails[];
   public vpcs: IVpc[];
   public regionalVpcs: { [region: string]: IVpc[] };
-  public allSecurityGroups: ISecurityGroup[];
+  public allSecurityGroups: ISecurityGroupsByAccountSourceData;
```
Comment (christopherthielen) [inline app/scripts/modules/amazon/src/securityGroup/configure/ingressRuleGroupSelector.component.ts]: woah 👍

File: app/scripts/modules/amazon/src/aws.settings.ts [hunk 2]
```diff
@@ -6,38 +6,41 @@ export interface IClassicLaunchWhitelist {
 }
 
 export interface IAWSProviderSettings extends IProviderSettings {
+  classicLaunchLockout?: number;
```
Comment (christopherthielen) [inline app/scripts/modules/amazon/src/aws.settings.ts]: did you sort this manually?

File: app/scripts/modules/amazon/src/securityGroup/configure/ingressRuleGroupSelector.component.ts [hunk 3]
```diff
@@ -116,7 +122,7 @@ class IngressRuleSelectorController implements IController {
         regionalVpcId = regionalVpc ? regionalVpc.id : undefined;
       }
 
-      const regionalGroupNames = get(this.allSecurityGroups, [account, 'aws', region].join('.'), [])
+      const regionalGroupNames = (this.allSecurityGroups?.[account].aws?.[region] ?? [])
```
Comment (christopherthielen) [inline app/scripts/modules/amazon/src/securityGroup/configure/ingressRuleGroupSelector.component.ts]:
```suggestion
      const regionalGroupNames = (this.allSecurityGroups?.[account]?.aws?.[region] ?? [])
```
?

File: app/scripts/modules/amazon/src/aws.settings.ts (hunk 2 above)
Comment (anotherchrisberry) [reply app/scripts/modules/amazon/src/aws.settings.ts]: yes

File: app/scripts/modules/core/src/artifact/expectedArtifact.service.ts [hunk 4]
```diff
@@ -39,7 +38,7 @@ export class ExpectedArtifactService {
     // is that we don't want to include an empty string in the artifact list, so omit any falsey values.
     return (stageContext: IExecutionContext) =>
       fields
-        .map((field): string => get(stageContext, field))
```
Comment (christopherthielen) [inline app/scripts/modules/core/src/artifact/expectedArtifact.service.ts]: 😮

File: app/scripts/modules/core/src/pipeline/config/PipelineRegistry.ts [hunk 5]
```diff
@@ -319,10 +317,10 @@ export class PipelineRegistry {
...
   private static resolveCloudProvider(stage: IStage): string {
     return (
-      stage.cloudProvider ||
-      stage.cloudProviderType ||
-      get(stage, ['context', 'cloudProvider']) ||
-      get(stage, ['context', 'cloudProviderType']) ||
+      stage.cloudProvider ??
+      stage.cloudProviderType ??
+      stage.context?.cloudProvider ??
+      stage?.context?.cloudProviderType ??
```
Comment (christopherthielen) [inline app/scripts/modules/core/src/pipeline/config/PipelineRegistry.ts]:
```suggestion
      stage.context?.cloudProviderType ??
```

stage can't be nullish at this point and the difference between the lines made me look thrice

File: app/scripts/modules/core/src/artifact/expectedArtifact.service.ts (hunk 4 above)
Comment (anotherchrisberry) [reply app/scripts/modules/core/src/artifact/expectedArtifact.service.ts]: oh, actually, yes, this is bad - there is one place where there's a dot-nested reference so we should use `get` here

**Comment by anotherchrisberry** [comment]
@louisjimenez yeah, it should all be transpiled down. I think both nullish coalescing and optional chaining are still drafts/proposals, so TS should transpile them indefinitely.

I didn't touch the JS files where we're using `get` and I'm not sure it's worth doing, since we don't get the benefits of type checking.

**Comment by louisjimenez** [comment]
Hmmm, I'm not seeing the transpiling in the bundle when doing a prod build. Agreed, that this feature has the most value for the ts files.

**Pull request merged**